### PR TITLE
feat: unit tests for link checking

### LINKED ISSUE #9453: Phishing detection
### Is your feature request related to a problem? Please describe.

As a Nextcloud Mail user, I want the app to help me detect phishing emails. Some of these phishing emails are not easy to spot.

### Describe the solution you'd like

Run some sanity checks on the email (to be brainstormed) and
1. Show a warning
2. Disable external links
3. Offer to move the message to spam

### Work packages
- [x] Research, brainstorming and goal definition
  - [x] Compare sender with sender label
  - [x] Trusted domains
  - [x] DKIM
  - [x] BIMI https://github.com/nextcloud/mail/pull/5784
- [x] Implementation
   - [x]  anchor tag text vs link checking #9994 
   - [x] Highlighting messages with a different reply-to #9610
   - [x] check full name address against contacts #9610
- [x] #9891
- [ ] Unit test for link checking https://github.com/nextcloud/mail/pull/10075
- [x] Admin docs if applicable (Not applicable) 
- [x] https://github.com/nextcloud/documentation/pull/12152

### Follow-up 

- [ ] Add a setting to disable HTML emails by default  
- [ ] Add quick actions (example: add email address to contact)
- [ ] Check emails for selectively applied CSS on forwards
context : https://www.schneier.com/blog/archives/2024/04/security-vulnerability-of-html-emails.html 

### Suggestion 
- [ ] Add a hint if a mail comes from an address that is neither in "contacts" nor has send any mails before. @kaffeeringe 

### Out of scope 
- resolving domain names
- Build a process to integrate against configurable Lists
- SPF 
- Preconfigured trust custom header tags 

### Describe alternatives you've considered

N/a

### Additional context

https://github.com/nextcloud/mail/issues/7864

### PR DESCRIPTION
**Pull request by hamza221**
requires  #10073

Fixes #9453

### CODE REVIEW

File: tests/Unit/Service/Phishing/LinkCheckTest.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+use ChristophWurst\Nextcloud\Testing\TestCase;
+
+use OCA\Mail\Service\PhishingDetection\LinkCheck;
+use OCP\IL10N;
+
+use PHPUnit\Framework\MockObject\MockObject;
+
+class LinkCheckTest extends TestCase {
+
+	private IL10N|MockObject $l10n;
```
Comment (kesselb) [inline tests/Unit/Service/Phishing/LinkCheckTest.php:18]:
```suggestion
	private IL10N&MockObject $l10n;
```

File: tests/Unit/Service/Phishing/LinkCheckTest.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
+<?php
+/**
```
Comment (kesselb) [inline tests/Unit/Service/Phishing/LinkCheckTest.php:2]:
```suggestion
<?php

declare(strict_types=1);

/**
```

File: tests/Unit/Service/Phishing/LinkCheckTest.php [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+use OCA\Mail\Service\PhishingDetection\LinkCheck;
+use OCP\IL10N;
+
+use PHPUnit\Framework\MockObject\MockObject;
+
+class LinkCheckTest extends TestCase {
+
+	private IL10N|MockObject $l10n;
+	private LinkCheck|MockObject $service;
```
Comment (kesselb) [inline tests/Unit/Service/Phishing/LinkCheckTest.php:19]:
```suggestion
	private LinkCheck&MockObject $service;
```

**Comment by kesselb** [comment]
Merge this one into the other branch? ;) The tests are failing, because the fix is not in, yet.

**Comment by hamza221** [comment]
> Merge this one into the other branch? ;) The tests are failing, because the fix is not in, yet.
I can rebase and merge to main later

File: tests/Unit/Service/Phishing/LinkCheckTest.php (hunk 3 above)
Comment (kesselb) [inline tests/Unit/Service/Phishing/LinkCheckTest.php:19]:
```suggestion
	private LinkCheck $service;
```

Comment (ChristophWurst) [reply tests/Unit/Service/Phishing/LinkCheckTest.php:21]: @hamza221 this is not done

Comment (ChristophWurst) [reply tests/Unit/Service/Phishing/LinkCheckTest.php:21]: @hamza221 this is not done

File: tests/Unit/Service/Phishing/LinkCheckTest.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+		$this->l10n = $this->createMock(IL10N::class);
+		$this->service = new LinkCheck($this->l10n);
+	}
+
+	public function testNakedAddressPass(): void {
+		$htmlMessage = '<html><body><a href="https://nextcloud.com/">nextcloud.com</a></p></body></html>';
+
+		$result = $this->service->run($htmlMessage);
+		$this->assertFalse($result->isPhishing());
+
```
Comment (ChristophWurst) [inline tests/Unit/Service/Phishing/LinkCheckTest.php:34]:
```suggestion
```

File: tests/Unit/Service/Phishing/LinkCheckTest.php [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	public function testCompleteAddressPass(): void {
+		$htmlMessage = '<html><body><a href="https://nextcloud.com/">https://nextcloud.com/</a></body></html>';
+
+		$result = $this->service->run($htmlMessage);
+		$this->assertFalse($result->isPhishing());
+
```
Comment (ChristophWurst) [inline tests/Unit/Service/Phishing/LinkCheckTest.php:42]: same

File: tests/Unit/Service/Phishing/LinkCheckTest.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+	public function testDeepAddressPass(): void {
+		$htmlMessage = '<html><body><a href="https://nextcloud.com/"><span class="text-big" style="color:hsl(0,75%,60%);font-family: Courier, monospace;"><i><strong>nextcloud.com</strong></i></span></a></body></html>';
+		
+		$result = $this->service->run($htmlMessage);
+		$this->assertFalse($result->isPhishing());
+
+	}
+
+
+
```
Comment (ChristophWurst) [inline tests/Unit/Service/Phishing/LinkCheckTest.php:64]: this will bloat the git repository by 3 bytes -> :fire:

**Review by ChristophWurst — COMMENTED** [review]
Whitespaces in test are a bit random

One way to structure them is the A-A-A pattern

* Arrange block
* *Blank line*
* Act block
* *Blank line*
* Assert block

This would fit nicely for the tests here.

**Review by ChristophWurst — APPROVED** [review]
Noice

**Review by ChristophWurst — COMMENTED** [review]
Ugh one more thing: make the commit a `test:`. This is not a user-facing feature :grimacing:

**Pull request merged**